core/frontend/src/extension/ExtensionImpl.ts

```diff
@@ -0,0 +1,46 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Extensions
+ */
+
+import { ToolType } from "../tools/Tool";
+import { IModelApp } from "../IModelApp";
+import { CommonToolbarItem, StageUsage, ToolbarItemUtilities, ToolbarOrientation, ToolbarUsage, UiItemsManager, UiItemsProvider } from "@itwin/appui-abstract";
+
+/** @beta */
+export class ToolProvider implements UiItemsProvider {
+  public readonly id;
+  private _toolId = "";
+  private _toolIcon;
+  private _toolLabel;
+
+  public constructor(tool: ToolType) {
+    this.id = `ToolProvider:${tool.toolId}`;
+    this._toolId = tool.toolId;
+    this._toolIcon = tool.iconSpec;
+    this._toolLabel = tool.description;
+  }
+
+  public provideToolbarButtonItems(_stageId: string, stageUsage: string, toolbarUsage: ToolbarUsage, toolbarOrientation: ToolbarOrientation): CommonToolbarItem[] {
+    const toolbarItem = ToolbarItemUtilities.createActionButton(this._toolId, 0, this._toolIcon, this._toolLabel, async () => { await IModelApp.tools.run(this._toolId); });
+    return stageUsage === StageUsage.General && toolbarUsage === ToolbarUsage.ContentManipulation && toolbarOrientation === ToolbarOrientation.Vertical ? [toolbarItem] : [];
```

Comment:
I understand that this is V1 but this is very restrictive (basically can only add toolbar buttons to a single location)

Reply:
Yes, this is one of the first things we should tackle after the initial release. I'll make sure to put that on our list.

Reply:
It's fine for v1. Getting the framework out is most important

Reply:
Its even more restrictive, its heavily tied with AppUI which idk is something we want to tie ourselves down to

I think we should think how extensions would plug into something not based on AppUi like display-test-app

Reply:
@aruniverse Yes - this is exactly why display-test-app does not use our UI packages - to try to prevent implicit dependencies on them from creeping in elsewhere.

Reply:
it's an "interesting" discussion. We certainly don't want to be tied to React, but the "provider" concept in general can probably be reused/re-purposed. The fact that we are allowing for extending UI that does not exist is not a trivial problem to solve

Reply:
True, I think we should definitely follow up on this and it's going to be a lengthy discussion. However, in terms of this PR are we okay with just hard-coding this how it is today?

As far as I know we did not expose anything explicitly AppUI related within the API and it's all under the hood to be potentially refactored later. If not, we should clear that up before merging.

Reply:
this is good as is